Mathlib/Combinatorics/SetFamily/Shatter.lean

```diff
@@ -0,0 +1,196 @@
+  rw [←card_union_add_card_inter, ←hℬ, ←card_disjoint_union]
+  swap
+  · simp only [disjoint_left, mem_union, mem_shatterer, mem_image, not_exists, not_and]
+    rintro _ (hs | hs) s - rfl
+    · exact aux (fun t ht ↦ (mem_memberSubfamily.1 ht).2) hs <| mem_insert_self _ _
+    · exact aux (fun t ht ↦ (mem_nonMemberSubfamily.1 ht).2) hs <| mem_insert_self _ _
+  refine card_mono <| union_subset (union_subset ?_ <| shatterer_mono <| filter_subset _ _) ?_
+  · simp only [subset_iff, mem_shatterer]
+    rintro s hs t ht
+    obtain ⟨u, hu, rfl⟩ := hs ht
+    rw [mem_memberSubfamily] at hu
+    refine ⟨insert a u, hu.1, inter_insert_of_not_mem fun ha ↦ ?_⟩
+    obtain ⟨v, hv, hsv⟩ := hs.exists_inter_eq_singleton ha
+    rw [mem_memberSubfamily] at hv
+    rw [←singleton_subset_iff (a := a), ←hsv] at hv
+    exact hv.2 <| inter_subset_right _ _
+  · refine forall_image.2 fun s hs ↦ mem_shatterer.2 fun t ht ↦ ?_
+    simp only [mem_inter, mem_shatterer] at hs
+    rw [subset_insert_iff] at ht
+    by_cases ha : a ∈ t
+    · obtain ⟨u, hu, hsu⟩ := hs.1 ht
+      rw [mem_memberSubfamily] at hu
+      refine ⟨_, hu.1, ?_⟩
+      rw [←insert_inter_distrib, hsu, insert_erase ha]
+    · obtain ⟨u, hu, hsu⟩ := hs.2 ht
+      rw [mem_nonMemberSubfamily] at hu
+      refine ⟨_, hu.1, ?_⟩
+      rwa [insert_inter_of_not_mem hu.2, hsu, erase_eq_self]
+
+lemma Shatter.of_compression (hs : (𝓓 a 𝒜).Shatter s) : 𝒜.Shatter s := by
+  intros t ht
+  obtain ⟨u, hu, rfl⟩ := hs ht
+  rw [Down.mem_compression] at hu
+  obtain hu | hu := hu
+  · exact ⟨u, hu.1, rfl⟩
+  by_cases ha : a ∈ s
+  · obtain ⟨v, hv, hsv⟩ := hs <| insert_subset ha ht
+    rw [Down.mem_compression] at hv
+    obtain hv | hv := hv
+    · refine ⟨erase v a, hv.2, ?_⟩
+      rw [inter_erase, hsv, erase_insert]
+      rintro ha
+      rw [insert_eq_self.2 (mem_inter.1 ha).2] at hu
+      exact hu.1 hu.2
+    rw [insert_eq_self.2 <| inter_subset_right s _ ?_] at hv
+    cases hv.1 hv.2
+    rw [hsv]
+    exact mem_insert_self _ _
+  · refine ⟨insert a u, hu.2, ?_⟩
+    rw [inter_insert_of_not_mem ha]
+
+lemma shatterer_compress_subset_shatterer (a : α) (𝒜 : Finset (Finset α)) :
+    (𝓓 a 𝒜).shatterer ⊆ 𝒜.shatterer := by
+  simp only [subset_iff, mem_shatterer]; exact fun s hs ↦ hs.of_compression
+
+/-! ### Vapnik-Chervonenkis dimension -/
+
+/-- The Vapnik-Chervonenkis dimension of a set family is the maximal size of a set it shatters. -/
+def vcDim (𝒜 : Finset (Finset α)) : ℕ := 𝒜.shatterer.sup card
+
+lemma Shatter.card_le_vcDim (hs : 𝒜.Shatter s) : s.card ≤ 𝒜.vcDim := le_sup <| mem_shatterer.2 hs
+
+/-- Down-compressing decreases the VC-dimension. -/
+lemma vcDim_compress_le (a : α) (𝒜 : Finset (Finset α)) : (𝓓 a 𝒜).vcDim ≤ 𝒜.vcDim :=
+  sup_mono <| shatterer_compress_subset_shatterer _ _
+
+/-- The **Sauer-Shelah lemma**. -/
+lemma card_shatterer_le_sum_vcDim [Fintype α] :
+    𝒜.shatterer.card ≤ ∑ k in Iic 𝒜.vcDim, (Fintype.card α).choose k := by
+  simp_rw [←card_univ, ←card_powersetLen]
```

Comment:
#7667 renamed `powersetLen` to `powersetCard`, so you'll have to merge master manually.